Add details drawer to Grid View

bbovenzi opened the pull request:
Add a details drawer to the right side of the grid view to quickly see details and actions for the DAG / Dag Run / Task Instance. This replaces the modal when clicking on a dag run or task instance. (Modals still exist for graph and gantt views)

Benefits:
- Simplify tooltips to only show basic at-a-glance info, and rely on the drawer for more advanced details
- Show details in context of what run or task you are looking at
- Dag Run and Task Instance actions in context of what Task/Run you are changing
- Run and Task actions is a seamless experience without multiple page redirects. And one can immediately see changes take place.

Next Up:
- The side drawer for a mapped task  will include a table of all of its mapped instances with their respective actions
- A better DAG details side drawer. In this PR, it is a table of some basic details, but I would like for it to be an overview of your DAG and the runs/tasks that are currently being displayed in the grid view

Closes #19938
Closes #21428
Related #20382
Closes #19674

Other notes:
Python isn't my strong suit so any help with the changes I had to make to the webserver (`views.py` and `utils.py`) would be greatly appreciated.

Dag Details:
<img width="991" alt="Screen Shot 2022-03-17 at 12 40 31 PM" src="https://user-images.githubusercontent.com/4600967/158852846-d5894305-b27d-4587-964d-d66ee06e6764.png">


Run Details:
<img width="998" alt="Screen Shot 2022-03-17 at 12 37 11 PM" src="https://user-images.githubusercontent.com/4600967/158852707-109fbf57-aa3f-4c67-b283-aa3803adc258.png">


Task Instance Details:
<img width="998" alt="Screen Shot 2022-03-17 at 12 37 02 PM" src="https://user-images.githubusercontent.com/4600967/158852657-e2c561cb-ce68-4681-b290-381b24a9a836.png">


Navigating between details:
![Mar-17-2022 12-56-37](https://user-images.githubusercontent.com/4600967/158853607-61fd6108-fd28-46b6-8255-e09331ee0081.gif)


Scrolling:
![Mar-17-2022 12-46-07](https://user-images.githubusercontent.com/4600967/158852919-e9786e28-ff7d-41a6-91d3-12aa8835f3b1.gif)


Marking a Task Failed:
![Mar-17-2022 12-51-06](https://user-images.githubusercontent.com/4600967/158853042-67b191e4-d754-43af-88e3-2fecee26e215.gif)


Clearing a DAG Run:
![Mar-17-2022 12-51-57](https://user-images.githubusercontent.com/4600967/158853227-4105d6de-7fb0-4b55-b37c-a4039bfa7f40.gif)

Changing Timezone:
![Mar-17-2022 21-03-22](https://user-images.githubusercontent.com/4600967/158917936-4604529f-18ca-4d5e-bd37-2973c1c0af6c.gif)



---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

bbovenzi committed:
make UI and tree work with mapped tasks

basic slide drawer

reformat grid background colors

improve rendering and add selected dag run

fix hover and extra prop

switch from drawer to details section

add tooltip info to details

use API

make side panel collapsible, useTasks,

dag run actions

dag run actions w/ react-query

task instance links

task actions

remove modals

adjust panel width and use status color

minor details styling

add duration to tooltips

add last scheduling decision and fix tests

ashb commented on airflow/www/static/js/tree/InstanceTooltip.jsx:
```diff
@@ -78,86 +76,27 @@ const InstanceTooltip = ({
+        Duration:
         {' '}
-        {endDate && formatDateTime(endDate)}
+        {formatDuration(duration || getDuration(startDate, endDate))}
```
What does this show when the task is still in progress?

Do we not show end date anymore?

bbovenzi replied:
`endDate` is in the details but not in the tooltip.
When in progress it will use `Date.now()` instead of enddate

bbovenzi committed:
move ref and selection to providers

bbovenzi committed:
fix test with mock providers

bbovenzi committed:
update TI and DR buttons

bbovenzi committed:
download logs and external logs

bbovenzi committed:
add extra links to TI details

bbovenzi committed:
download log bug fixes

bbovenzi committed:
fix extra links, hide local TZ if UTC,

bbovenzi committed:
confirm mark task failed/success

bbovenzi committed:
Update confirm modals for runs and tasks

- async/await on mutations instead of useeffect
- add confirmation for run actions

bbovenzi committed:
Fix dialog scrolling

bbovenzi commented on airflow/www/static/js/dag.js:
```diff
@@ -227,22 +227,6 @@ export function callModal(t, d, extraLinks, tryNumbers, sd, drID) {
   }
 }
 
-export function callModalDag(dag) {
```
We only called this in the grid view, so we can fully replace it

bbovenzi commented on airflow/www/templates/airflow/dag.html:
```diff
@@ -375,70 +380,6 @@ <h4>Task Actions</h4>
       </div>
     </div>
   </div>
-  <!-- Modal for DAG run -->
```
This is now all in React instead.

bbovenzi committed:
Code cleanup and fix task clear

bbovenzi committed:
Fix task/run label, dialog focus, dag details overflow, panel open/close

bbovenzi committed:
Add timezone provider

bbovenzi committed:
Fix TimezoneEvent import

bbovenzi committed:
Improve button UX

- Remove details panel title
- Add button to reset root
- Make "More Details" buttons more specific
- Specify timezone as DAG timezone

bbovenzi committed:
autorefresh dag run details

bbovenzi committed:
auto-refresh task instance details

ryanahamilton commented on airflow/www/static/js/tree/StatusBox.jsx:
```diff
@@ -24,55 +24,76 @@ import { isEqual } from 'lodash';
+  const { selected, onSelect } = useSelection();
+  const { runId, taskId } = instance;
+  const { colors } = useTheme();
+  const hoverBlue = `${colors.blue[100]}50`;
```
What's this translating to? Is it just suffixing the `50` to the hex value? e.g. `#BEE3F850`?

bbovenzi replied:
Yes, the `50` is the transparency value like you were using `rgba()` instead of `rgb()`

bbovenzi replied:
For reference: https://caniuse.com/css-rrggbbaa

ryanahamilton commented on airflow/www/static/js/tree/StatusBox.jsx:
```diff
@@ -24,55 +24,76 @@ import { isEqual } from 'lodash';
-        data-testid="task-instance"
-      />
-    </Tooltip>
+    <>
```
Why the added fragment? Doesn't appear necessary given `<Tooltip />` is the only child.

ryanahamilton commented on airflow/www/static/js/tree/details/Header.jsx:
```diff
@@ -0,0 +1,95 @@
+
+  return (
+    <Breadcrumb color="gray.300">
+      <BreadcrumbItem isCurrentPage={!runId && !taskId} mt="15px">
```
Could the `mt="15px"` on each of these `BreadcrumbItem` be moved up to the parent `Breadcrumb`? Could we just use Chakra spacing instead? e.g.  `mt={4}`

ryanahamilton commented on airflow/www/static/js/tree/details/content/Dag.jsx:
```diff
@@ -0,0 +1,137 @@
+  });
+
+  return (
+    <Box>
```
Is this `Box` doing anything or could it be a fragment instead?

ryanahamilton commented on airflow/www/static/js/tree/details/content/dagRun/ClearRun.jsx:
```diff
@@ -0,0 +1,65 @@
+        isOpen={isOpen}
+        onClose={onClose}
+        onConfirm={onConfirm}
+        description="Here's the list of task instances you are about to clear:"
```
I didn't see this copy in the PR GIF, but even without the context, it seems like the copy could be simplified:
```suggestion
        description="Task instances you are about to clear:"
```

ryanahamilton commented on airflow/www/static/js/tree/details/content/dagRun/MarkFailedRun.jsx:
```diff
@@ -0,0 +1,61 @@
+        isOpen={isOpen}
+        onClose={onClose}
+        onConfirm={onConfirm}
+        description="Here's the list of task instances you are about to mark as failed:"
```
Similar copy suggestion here
```suggestion
        description="Task instances you are about to mark as failed:"
```

ryanahamilton commented on airflow/www/static/js/tree/details/content/dagRun/index.jsx:
```diff
@@ -0,0 +1,140 @@
+import { MdPlayArrow, MdOutlineAccountTree } from 'react-icons/md';
+
+import { SimpleStatus } from '../../../StatusBox';
+import { formatDuration } from '../../../../datetime_utils';
```
Not for this PR, but we might want to add some path aliasing so we can avoid all of the relative path imports.

bbovenzi replied:
Absolutely. It pained me to write all of these relative path imports.

ryanahamilton commented on airflow/www/static/js/tree/details/content/dagRun/QueueRun.jsx:
```diff
@@ -0,0 +1,74 @@
+        isOpen={isOpen}
+        onClose={onClose}
+        onConfirm={onConfirm}
+        description="Here's the list of task instances you are about to queue:"
```
Similar copy suggestion here
```suggestion
        description="Task instances you are about to queue:"
```

bbovenzi committed:
revert useTreeData changes

None of these changes were relevant to this PR. Better to be done separately.

ryanahamilton commented on airflow/www/static/js/tree/details/content/taskInstance/ExtraLinks.jsx:
```diff
@@ -0,0 +1,64 @@
+  });
+
+  if (!links.length) return null;
+  const checkIfExternal = (url) => /^(?:[a-z]+:)?\/\//.test(url);
```
Nit, the function naming could make it more explicit that returns a boolean
```suggestion
  const isExternal = (url) => /^(?:[a-z]+:)?\/\//.test(url);
```

ryanahamilton commented on airflow/www/static/js/tree/details/content/taskInstance/taskActions/Clear.jsx:
```diff
@@ -0,0 +1,127 @@
+  const [affectedTasks, setAffectedTasks] = useState([]);
+
+  // Options check/unchecked
+  const { isOpen: past, onToggle: onTogglePast } = useDisclosure();
```
Seems a little unorthodox to use `useDisclosure` to manage these boolean state values, no?

bbovenzi replied:
Fair enough. Updated to `useState(false)`

bbovenzi committed:
Address PR feedback

- useState vs useDisclosure
- Remove extraneous elements
- Copy changes
- Wire up params for runTask
- Breadcrumb padding

bbovenzi committed:
Handle task/run action sideeffects by separating autorefresh and treeData hooks

uranusjr commented on airflow/www/views.py:
```diff
@@ -1718,6 +1730,7 @@ def run(self, session=None):
         dag_run_id = request.form.get('dag_run_id')
         map_index = request.args.get('map_index', -1, type=int)
         origin = get_safe_url(request.form.get('origin'))
+        return_as_json = request.form.get('return_as_json') == "true"
```
I wonder if it’s better to use HTTP header for this. We could use `request.content_type` (automatically return JSON for `application/json` and redirect otherwise), or the conventional `X-Requested-With`.

bbovenzi replied:
Good call. I added a check for `request.headers['Accept'] == 'application/json'` instead of a param

uranusjr commented on airflow/www/views.py:
```diff
@@ -1961,24 +1976,53 @@ def _clear_dag_tis(
+        else:
+            details = [str(t) for t in tis]
+
+            return htmlsafe_json_dumps(details, separators=(',', ':'))
```
Is this the same as `_clear_dag_tis` but without `flash` and with different return values? We can probably combine them if that’s the case.

bbovenzi replied:
Good idea. I realized that we could also combine the two new confirmation endpoints back into the originals.

bbovenzi committed:
Clean up views.py endpoints

- Pass 'Accept' headers for json returns
- Consolidate more endpoints to return json or redirect

uranusjr commented on airflow/www/views.py:
```diff
@@ -527,6 +527,21 @@ def get_task_stats_from_query(qry):
     return data
 
 
+def redirect_or_json(origin, return_as_json, msg, status=""):
```
I think we can remove `return_as_json` entirely and check `request.headers['Accept'] == 'application/json'` in this function directly

bbovenzi replied:
Yeah, I didn't like that variable name. Updated.

uranusjr replied:
Actually you don’t even need to pass in `request`; the variable is magical in Flask and will automatically refer to the correct thing when it’s called in a view function.

bbovenzi replied:
TIL

bbovenzi replied:
Hmm. That seemed to cause some testing errors with `KeyError: 'HTTP_ACCEPT'`. Would you have any idea on how to address that?

bbovenzi committed:
pass request as arg

bbovenzi committed:
remove request as arg

uranusjr reviewed (approved):
For the Python part

uranusjr committed:
Anticipate when the 'Accept' header is not present

bbovenzi committed:
Fix argument count errors

ashb commented on airflow/www/static/js/tree/ResetRoot.jsx:
```diff
@@ -0,0 +1,55 @@
+    const params = new URLSearchParams();
+    if (numRuns) params.set('num_runs', numRuns);
+    if (baseDate) params.set('base_date', baseDate);
+    rootLink = `/dags/${dagId}/grid?${params.toString()}`;
```
We can't hard code the URL like this - airflow can be rooted/deployed to a location other than `/`.

In other places we put the URL in a meta tag from the jinja templates for this purpose.

bbovenzi replied:
All urls in react should be updated now.

ashb commented on airflow/www/static/js/tree/api/index.js:
```diff
@@ -0,0 +1,57 @@
+axios.defaults.headers.common.Accept = 'application/json';
+
+export {
+  useDag,
```
Why the `use` prefix on all these functions/nodules? It looks slightly odd to me

bbovenzi replied:
They are hooks, which are a particular type of function in React. https://reactjs.org/docs/hooks-intro.html

>A custom Hook is a JavaScript function whose name starts with ”use” and that may call other Hooks.

ashb commented on airflow/www/static/js/tree/api/useClearTask.js:
```diff
@@ -0,0 +1,77 @@
+        only_failed: failed,
+      }).toString();
+
+      return axios.post('/clear', params, {
```
Same comment here about hard-coding URLs

bbovenzi committed:
Replace hard coded urls

bbovenzi committed:
Replace hard coded urls in react components

bbovenzi committed:
Update filter upstream link

bbovenzi commented on airflow/www/templates/airflow/calendar.html:
```diff
@@ -22,7 +22,6 @@
 
 {% block head_css %}
   {{ super() }}
-  <meta name="grid_url" content="{{ url_for('Airflow.grid', dag_id=dag.dag_id) }}">
```
Adding `grid_url` to `dag.html` so this is now redundant.

bbovenzi committed:
Split TaskInstance details component

bbovenzi committed:
Fix undefined variables in tests

bbovenzi commented on airflow/www/views.py:
```diff
@@ -2554,6 +2552,20 @@ def grid(self, dag_id, session=None):
 
+        dag_details_url = ''
+        tasks_url = ''
+        # This is a workround to use try/except if url_for() throws an error
```
During tests `url_for` was throwing an error, so I moved the logic into `views.py` but I'm all ears on a better way to handle this.
In the future, it would be nice to have a sort of `url_for(RestAPI)` function.

jedcunningham replied:
The api init step is skipped in those test. Try this instead:

```
--- a/tests/www/views/test_views_log.py
+++ b/tests/www/views/test_views_log.py
@@ -56,7 +56,12 @@ def backup_modules():
 @pytest.fixture(scope="module")
 def log_app(backup_modules):
     @dont_initialize_flask_app_submodules(
-        skip_all_except=["init_appbuilder", "init_jinja_globals", "init_appbuilder_views"]
+        skip_all_except=[
+            "init_appbuilder",
+            "init_jinja_globals",
+            "init_appbuilder_views",
+            "init_api_connexion",
+        ]
     )
     @conf_vars({('logging', 'logging_config_class'): 'airflow_local_settings.LOGGING_CONFIG'})
     def factory():
```

bbovenzi committed:
init_api_connexion in tests

- add readme
-  rename context providers to avoid confusion with Airflow Providers

bbovenzi committed:
Fix url params, hide last item breadcrumb links

bbovenzi committed:
Update task run failed copy

bbovenzi committed:
Fix taskinstance/list buttons
